### PR TITLE
[MNG-6511] - Optional project selection

### PR DESCRIPTION
**Pull request by MartinKanters**
Relevant JIRA issue: https://issues.apache.org/jira/browse/MNG-6511

### CODE REVIEW

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng6511OptionalProjectSelectionTest.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    public void testSelectNonExistingOptionalProfile() throws VerificationException
+    {
+        newVerifier( testDir.getAbsolutePath() ).executeGoal( "clean" );
+
+        Verifier verifier = newVerifier( testDir.getAbsolutePath() );
+        verifier.setLogFileName( "log-select-non-existing.txt" );
+        verifier.addCliOption( "-pl ?non-existing-module" );
+        verifier.executeGoal( "validate" );
+        verifier.verifyErrorFreeLog();
+        verifier.assertFilePresent( "existing-module/target/touch.txt" ); // existing-module should have been built.
```
Comment (michael-o) [inline core-it-suite/src/test/java/org/apache/maven/it/MavenITmng6511OptionalProjectSelectionTest.java:78]: How does this line relate to `non-exsisting-module`?

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng6511OptionalProjectSelectionTest.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    public void testDeselectNonExistingOptionalProfile() throws VerificationException
+    {
+        newVerifier( testDir.getAbsolutePath() ).executeGoal( "clean" );
+
+        Verifier verifier = newVerifier( testDir.getAbsolutePath() );
+        verifier.setLogFileName( "log-deselect-non-existing.txt" );
+        verifier.addCliOption( "-pl !?non-existing-module" );
+        verifier.executeGoal( "validate" );
+        verifier.verifyErrorFreeLog();
+        verifier.assertFilePresent( "existing-module/target/touch.txt" ); // existing-module should have been built.
```
Comment (michael-o) [inline core-it-suite/src/test/java/org/apache/maven/it/MavenITmng6511OptionalProjectSelectionTest.java:102]: Same here.

File: core-it-suite/src/test/resources/mng-6511-optional-project-selection/pom.xml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  </properties>
+
+  <modules>
+    <module>existing-module</module>
+  </modules>
+
+  <build>
+    <pluginManagement>
+      <!-- maven-resources-plugin, maven-compiler-plugin -->
+      <plugins>
```
Comment (michael-o) [inline core-it-suite/src/test/resources/mng-6511-optional-project-selection/pom.xml:51]: Did you check whether these are downloaded via bootstrap IT? If not, use the same versions as bundled with Maven 4.

**Review by michael-o — CHANGES_REQUESTED** [review]

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng6511OptionalProjectSelectionTest.java (hunk 1 above)
Comment (MartinKanters) [reply core-it-suite/src/test/java/org/apache/maven/it/MavenITmng6511OptionalProjectSelectionTest.java:78]: Well, our idea was the following, if someone tries to select a module optionally and it does not exist, the build should just build everything. In other words `mvn validate -pl ?non-existing` has the same outcome as `mvn validate`.

File: core-it-suite/src/test/resources/mng-6511-optional-project-selection/pom.xml (hunk 3 above)
Comment (MartinKanters) [reply core-it-suite/src/test/resources/mng-6511-optional-project-selection/pom.xml:51]: Yep, [they are](https://github.com/apache/maven-integration-testing/blob/master/core-it-suite/src/test/resources/bootstrap/pom.xml#L61-L80) :)

File: core-it-suite/src/test/java/org/apache/maven/it/MavenITmng6511OptionalProjectSelectionTest.java (hunk 1 above)
Comment (michael-o) [reply core-it-suite/src/test/java/org/apache/maven/it/MavenITmng6511OptionalProjectSelectionTest.java:78]: Alright, I have noticed that in other tests too. Maybe this special case should be documented where we have profile selection in maven-site. By defiinition, non-existing will cause the entire build. Take it or leave it.

Comment (MartinKanters) [reply core-it-suite/src/test/java/org/apache/maven/it/MavenITmng6511OptionalProjectSelectionTest.java:78]: Yeah, sounds like a good idea! Updating the docs in maven-site is something we shouldn't forget, and especially this case.

Comment (michael-o) [reply core-it-suite/src/test/java/org/apache/maven/it/MavenITmng6511OptionalProjectSelectionTest.java:78]: Please create a MNGSITE ticket for this.

Comment (mthmulders) [reply core-it-suite/src/test/java/org/apache/maven/it/MavenITmng6511OptionalProjectSelectionTest.java:78]: See [MNGSITE-444](https://issues.apache.org/jira/browse/MNGSITE-444).

**Review by michael-o — APPROVED** [review]

**Pull request closed without merging**